File: csharp/tournament/Tournament.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Text;
using System.Linq;

public static class Tournament
{
    private const string ResultsHeader = "Team                           | MP |  W |  D |  L |  P\n";
    public static void Tally(Stream inStream, Stream outStream)
    {
        IDictionary<string, TeamResult> results = new Dictionary<string, TeamResult>();
        string tournamentResults = new StreamReader(inStream).ReadToEnd().Trim();

        foreach (string matchResult in tournamentResults.Split("\n", StringSplitOptions.RemoveEmptyEntries))
        {
            results = UpdateTeamResults(new ResultLine(matchResult), results);
        }

        outStream.Write(Encoding.UTF8.GetBytes(ShowResults(results)));
    }

    private static IDictionary<string, TeamResult> UpdateTeamResults(ResultLine resultLine, IDictionary<string, TeamResult> results)
    {
        switch (resultLine.Result)
        {
            case "draw":
                GetTeamResults(resultLine.Team1, results).RecordDraw();
                GetTeamResults(resultLine.Team2, results).RecordDraw();
                break;
            case "win":
                GetTeamResults(resultLine.Team1, results).RecordWin();
                GetTeamResults(resultLine.Team2, results).RecordLoss();
                break;
            case "loss":
                GetTeamResults(resultLine.Team1, results).RecordLoss();
                GetTeamResults(resultLine.Team2, results).RecordWin();
                break;
        }

        return results;
    }

    private static string ShowResults(IDictionary<string, TeamResult> results)
    {
        StringBuilder sb = new StringBuilder(ResultsHeader);

        var orderedResults = from result in results.Values
                             orderby result.Points descending, result.Name ascending
                             select result;

        foreach (var result in orderedResults)
        {
            sb.AppendLine(result.ToString());
        }

        return sb.ToString().Trim();
    }
    private static TeamResult GetTeamResults(string name, IDictionary<string, TeamResult> results)
    {
        if (results.TryGetValue(name, out var foundTeamResult))
        {
            return foundTeamResult;
        }

        var newTeam = new TeamResult(name);
        results.Add(name, newTeam);

        return newTeam;
    }
}

internal class ResultLine
{
    public string Result { get; }
    public string Team1 { get; }
    public string Team2 { get; }

    public ResultLine(string rawLine)
    {
        string[] resultFields = rawLine.Split(";");
        Result = resultFields[2];
        Team1 = resultFields[0];
        Team2 = resultFields[1];
    }
}
internal class TeamResult
{
    private int MatchesPlayed;
    private int Wins;
    private int Draws;
    private int Losses;
    public int Points { get; private set; }
    public string Name { get; }

    public TeamResult(string name) => Name = name;

    public void RecordWin()
    {
        MatchesPlayed++; Wins++; Points += 3;
    }
    public void RecordLoss()
    {
        MatchesPlayed++; Losses++; Points += 0;
    }
    public void RecordDraw()
    {
        MatchesPlayed++; Draws++; Points += 1;
    }

    public override string ToString() => $"{rightPadToWidth(Name, 30)} | {format(MatchesPlayed)} | {format(Wins)} | {format(Draws)} | {format(Losses)} | {format(Points)}";

    private string format(int value) => leftPadToWidth(value.ToString(), 2);

    private string leftPadToWidth(String value, int width) => new string(' ', width - value.Length) + value;

    private string rightPadToWidth(string value, int width) => value + new string(' ', width - value.Length);
}

File: solutions/ruby/tournament/7/tournament.rb
module Tournament
    class TeamResults
        def initialize(team)
            @team = team
            @played = @wins = @draws = @losses = @points = 0
        end

        def record_win()
            record_game()
            @wins += 1
            @points += 3

            self
        end

        def record_loss()
            record_game()
            @losses += 1

            self
        end

        def record_draw()
            record_game()
            @draws += 1
            @points += 1

            self
        end
        
        def to_s()
            "#{team.ljust(31)}| #{played.to_s.rjust(2)} | #{wins.to_s.rjust(2)} | #{draws.to_s.rjust(2)} | #{losses.to_s.rjust(2)} | #{points.to_s.rjust(2)}"
        end

        attr_reader :wins, :losses, :points, :draws, :played, :team

        private

        def record_game()
            @played += 1
        end

    end

    def self.tally(input)

        results = {}
        input.split("\n").each do 
            | line |
            team1, team2, result = line.split(";")
            team1_results = results.fetch(team1, TeamResults.new(team1))
            team2_results = results.fetch(team2, TeamResults.new(team2))
            
            case result
            when "win"
                results[team1] = team1_results.record_win
                results[team2] = team2_results.record_loss
            when "loss"
                results[team1] = team1_results.record_loss
                results[team2] = team2_results.record_win
            when "draw"
                results[team1] = team1_results.record_draw
                results[team2] = team2_results.record_draw
            end
        end

        table_order = results.sort_by { | _team,  result | [-result.points, result.team] }
        teams = table_order.map { | _team, results | "#{results.to_s}\n" }.join()
        
        "Team                           | MP |  W |  D |  L |  P\n" + teams
    end
end